server/src/main/java/org/opensearch/common/geo/GeoUtils.java

```diff
@@ -444,112 +448,201 @@ public static GeoPoint parseGeoPoint(XContentParser parser, GeoPoint point, fina
+        switch (parser.currentToken()) {
+            case START_OBJECT:
+                parseGeoPointObject(parser, point, ignoreZValue, effectivePoint);
+                break;
+            case START_ARRAY:
+                parseGeoPointArray(parser, point, ignoreZValue);
+                break;
+            case VALUE_STRING:
+                String val = parser.text();
+                point.resetFromString(val, ignoreZValue, effectivePoint);
+                break;
+            default:
+                throw new OpenSearchParseException("geo_point expected");
+        }
+        return point;
+    }
+
+    private static GeoPoint parseGeoPointObject(
+        final XContentParser parser,
+        final GeoPoint point,
+        final boolean ignoreZValue,
+        final GeoUtils.EffectivePoint effectivePoint
+    ) throws IOException {
+        try (XContentSubParser subParser = new XContentSubParser(parser)) {
+            if (subParser.nextToken() != XContentParser.Token.FIELD_NAME) {
+                throw new OpenSearchParseException(ERR_MSG_INVALID_TOKEN, subParser.currentToken());
+            }
+
+            String field = subParser.currentName();
+            if (LONGITUDE.equals(field) || LATITUDE.equals(field)) {
+                parseGeoPointObjectBasicFields(subParser, point);
+            } else if (GEOHASH.equals(field)) {
+                parseGeoHashFields(subParser, point, effectivePoint);
+            } else if (GEOJSON_TYPE.equals(field) || GEOJSON_COORDS.equals(field)) {
+                parseGeoJsonFields(subParser, point, ignoreZValue);
+            } else {
+                throw new OpenSearchParseException(ERR_MSG_INVALID_FIELDS);
+            }
+
+            if (subParser.nextToken() != XContentParser.Token.END_OBJECT) {
+                throw new OpenSearchParseException(ERR_MSG_INVALID_FIELDS);
+            }
+
+            return point;
+        }
+    }
+
+    private static GeoPoint parseGeoPointObjectBasicFields(final XContentParser parser, final GeoPoint point) throws IOException {
+        HashMap<String, Double> data = new HashMap<>();
+        for (int i = 0; i < 2; i++) {
+            if (i != 0) {
+                parser.nextToken();
+            }
+
+            if (parser.currentToken() != XContentParser.Token.FIELD_NAME) {
+                break;
+            }
+
+            String field = parser.currentName();
+            if (!LONGITUDE.equals(field) && !LATITUDE.equals(field)) {
+                throw new OpenSearchParseException(ERR_MSG_INVALID_FIELDS);
+            }
+            switch (parser.nextToken()) {
+                case VALUE_NUMBER:
+                case VALUE_STRING:
+                    try {
+                        data.put(field, parser.doubleValue(true));
+                    } catch (NumberFormatException e) {
+                        throw new OpenSearchParseException("[{}] and [{}] must be valid double values", e, LONGITUDE, LATITUDE);
```

Comment:
Three values passed while i see two [{}], How will the error message be displayed?

Reply:
`OpenSearchParseException(String msg, Throwable cause, Object... args)`